ADR/0032-decoupling-deployment.md

```diff
@@ -0,0 +1,211 @@
+  be organized into their own project: "the [deployment-target-operator]", or something like that.
+- [integration-service] should be modified to create and destroy [DeploymentTargetClaims] directly
+  instead of [Environment] resources. Users will be expected to provide integration test pipelines
+  that accept a kubeconfig and snapshot as they do now, but those pipelines should now take initial
+  steps to *deploy the instance of their application to be tested* to the deployment target provided
+  by [integration-service] using the provided kubeconfig.
+- We should promote [release-service] as the primary means to advertise to [renovatebot] that one or
+  more images have passed testing and are ready to be promoted to a particular environment (with
+  a lowercase "e") by way of image tags in a registry.
+
+### Out of scope
+
+Some other interesting ideas that are floating around these days, but which should be taken up in
+other ADRs, if we take them up at all:
+
+- Decoupling [integration-service] and [release-service] from a shared [Snapshot] API. They will
+  continue to share a [Snapshot] API as of this ADR.
+- Decoupling [integration-service] from the [application-service] APIs, like [Application] and
+  [Component]. It will still promote images to the "global candidate list" (on the [Component]
+  resources) and continue to use the list of [Components] to guide its construction of [Snapshots].
+
+### Use Case Descriptions
+
+**During onboarding**: whereas today when a user requests a new appstudio tier namespace, the tier
+template includes an [Environment] that the integration-service will promote to. Tomorrow, the
+appstudio tier template should no longer include an [Environment] which on its own will cause
+integration-service to _not_ trigger a deployment when testing completes. Instead, the appstudio
+tier template should include a [ReleasePlan] with a reference to the [push-to-registry]
+release pipeline. This new default [ReleasePlan] should carry parameters such that whenever
+a [Snapshot] is successfully tested, a [Release] is created that re-tags the images in build-time
+quay repositories with a tag like `:released` or `:validated` (name tbd). The
+[push-to-registry] pipeline can use the `appstudio-pipeline` service account in the user's
+dev workspace, which already has push access to the repository in question.
+
+```mermaid
+flowchart TD
+    NSTierTemplate --> |provides| ReleasePlan
+    A[fa:fa-code-commit Git Commit] --> |webhook| PipelinesAsCode
+    PipelinesAsCode --> |creates| build[Build PipelineRun]
+    build --> |triggers| integration-service
+    integration-service --> |asks| decision{Is there a ReleasePlan?}
+    ReleasePlan --> |influences| decision
+    decision --> |yes| Release
+    build --> |pushes the image| quay.io/redhat-user-workloads/tenant/component
+    Release --> |tags the image for release| quay.io/redhat-user-workloads/tenant/component
```

Comment:
It tags the images for release or mirrors them to another quey repo?
I'm a bit worried about the fact the released image will be pulled from a repo that is intended to store CI builds.

Reply:
I guess it _could_ tag the image (I assume we use the same SA that has permissions to modify the repo) for release but that will likely not be a big use case. I have always envisioned that the release process is to some other _self-managed_ lifecycle.

If it is not already clear, we need to make sure that we have it clearly documented that the images in the tenant/application/component repository are for CI use and _not_ to depend on their lifecycle for any other use cases.

Reply:
To clarify - I meant for this to be as you read it @gbenhaim: for the release-service pipeline to tag the image in the same repo where it was built. It does have the dangers you worry about.

I thought this would be preferable as a default ReleasePlan that is just "marking" which build in the CI quay repo is "latest". Perhaps the default tag name should be chosen to reflect that low-degree of trustworthiness (say just "latest" rather than "release" or "validated").

To graduate out of an onboarding use case, the user should either update their ReleasePlan to instruct their tenant release pipeline to release to a separately managed image repository, or update their ReleasePlan to associate with a ReleasePlanAdmission in a managed workspace.

Reply:
| To graduate out of an onboarding use case, the user should either update their ReleasePlan to instruct their tenant release pipeline to release to a separately managed image repository, or update their ReleasePlan to associate with a ReleasePlanAdmission in a managed workspace.

maybe we need to emphasis it here.